src/collections/maple_tree/mod.rs

```diff
@@ -370,159 +505,185 @@ where
+                    isNone = true;
                     break;
-                },
-            }   
+                }
+            }
         }
-        
 
         // if we didn't find any pivot where rangeLower < self.pivot[i]
         // then we either ran into a None value or since we don't check min or max
         // the range must be between pivots at the index's R_CAPACITY and R_CAPCITY-1
-        if !set{
-            if isNone{
-                return ops::Range{start: (index-1) as isize, end: index as isize};
-            }
-            else{
-
-                return ops::Range{start: (R_CAPACITY-1) as isize, end: R_CAPACITY as isize};
+        if !set {
+            if isNone {
+                return ops::Range {
+                    start: (index - 1) as isize,
+                    end: index as isize,
+                };
+            } else {
+                return ops::Range {
+                    start: (R_CAPACITY - 1) as isize,
+                    end: R_CAPACITY as isize,
+                };
             }
         }
 
         // now we know that the range starts at index rangeB-1 and we need to find
-        // where the range ends. 
-        for i in (rangeB-1)..R_CAPACITY{
-            match &self.pivot[i]{
-                M::Some(b2) => { 
-                    if rangeUpper <= b2{
-                        if rangeB == 1{
-                            return ops::Range{start: -1, end: i as isize};
+        // where the range ends.
+        for i in (rangeB - 1)..R_CAPACITY {
+            match &self.pivot[i] {
+                M::Some(b2) => {
+                    if rangeUpper <= b2 {
+                        if rangeB == 1 {
+                            return ops::Range {
+                                start: -1,
+                                end: i as isize,
+                            };
                         }
-                        return ops::Range{start: (rangeB-2) as isize, end: i as isize};
+                        return ops::Range {
+                            start: (rangeB - 2) as isize,
+                            end: i as isize,
+                        };
                     }
                 }
                 M::None => {
-                        return ops::Range{start: (rangeB-2) as isize, end: i as isize};
+                    return ops::Range {
+                        start: (rangeB - 2) as isize,
+                        end: i as isize,
+                    };
                 }
             }
         }
 
-        if rangeB == 1{
-            return ops::Range{start: -1, end: R_CAPACITY as isize};
-        }
-        else{
-            return ops::Range{start: (rangeB-2) as isize, end: R_CAPACITY as isize}; 
+        if rangeB == 1 {
+            return ops::Range {
+                start: -1,
+                end: R_CAPACITY as isize,
+            };
+        } else {
+            return ops::Range {
+                start: (rangeB - 2) as isize,
```

Comment:
usize not isize? Can there ever be a case where the value is negative? I guess I'm asking, why use signed values here?

Reply:
You also could use the checked subtract here on usize to guarantee we don't underflow/overflow, so we don't have to do it manually.

Reply:
I was using isize because the start of the range can be -1 to indicate the use of the min variable.
I guess end doesn't need to be isize but I just made it that to be consistent with start but I can just make it usize.

Reply:
If you use checked subtract https://doc.rust-lang.org/std/primitive.usize.html#method.checked_sub then you get Option<result>, where None means "-1" aka underflow, so you have a None instead. Then you can map that our to Min or whatever else :)

Reply:
It saves you all the casting which do have implict panics on failure btw.

Reply:
alright that sounds good, yeah the casting is a bit of a pain.
I didn't realise that the casting could fail?

Reply:
In rust if the value is not within the target type, it will panic. IE negative numbers as usize will cause a panic I think.

Reply:
I was having bugs where the value was being set to usize::MAX because it was getting casted from a -1 isize value so I don't think thats the case with usize and isize